lib/rubygems/commands/setup_command.rb

```diff
@@ -373,7 +373,7 @@ def install_default_bundler_gem
     bundler_spec = Gem::Specification.load(default_spec_path)
 
     Dir.entries(bundler_spec.gems_dir).
-      select {|default_gem| default_gem.start_with?("bundler-") }.
+      select {|default_gem| default_gem.match(/bundler-\d/) }.
```

Comment:
maybe `default_gem.match(/^bundler-#{Gem::Version::VERSION_PATTERN}$/)`?

Reply:
you didn't include the `^`, so `foo-bundler-1` could still be deleted

Reply:
`default_gem` returns `/full/path/of/gems/bundler-x.y.z` . I think it needs to `File.basename(default_gem).match(/.../)` .

I will revise it with changes of other PRs.